File: 2019/0918/exponentiation.java
import java.util.Scanner;

public class Main {

  static Scanner sc = new Scanner(System.in);

  public static void main(String[] args) {
    long n = sc.nextLong();
    long m = sc.nextLong();
    int mod = 10007;

    long result = 1;
    long multiply = n;
    while (m > 0) {
      if (m % 2 == 1)
        result *= multiply;
      result %= mod;
      multiply = ((multiply % mod) * (multiply % mod)) % mod;
      m /= 2;
    }
    System.out.print(result);
  }
}

File: 2019/0918/exponentiation.py
import sys

n, m = map(int, sys.stdin.readline().split())

def calcPow(n, m):
  result = 1
  multiply = n
  mod = 10007
  while m > 0:
    if m % 2 == 1:
      result *= multiply
      result %= mod;
    multiply = ((multiply%mod)*(multiply%mod))%mod;
    m = m//2
  return result

print(calcPow(n, m))